src/Bedrock.Framework.Experimental/Protocols/Kafka/Utils/BufferWriterExtensions.cs

```diff
@@ -0,0 +1,223 @@
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Text;
+
+namespace Bedrock.Framework.Protocols.Kafka.Utils
+{
+    public static class BufferWriterExtensions
+    {
+        public static void WriteByte(this IBufferWriter<byte> output, byte num)
+        {
+            var buffer = output.GetSpan(1);
+            buffer[0] = num;
+            output.Advance(1);
+        }
+
+        public static void WriteShort(this IBufferWriter<byte> output, short num)
+        {
+            var buffer = output.GetSpan(2);
+            BinaryPrimitives.WriteInt16BigEndian(buffer, num);
+            output.Advance(2);
+        }
+
+        public static void WriteUShort(this IBufferWriter<byte> output, ushort num)
+        {
+            var buffer = output.GetSpan(2);
+            BinaryPrimitives.WriteUInt16BigEndian(buffer, num);
+            output.Advance(2);
+        }
+
+        public static void WriteInt(this IBufferWriter<byte> output, int num)
+        {
+            var buffer = output.GetSpan(4);
+            BinaryPrimitives.WriteInt32BigEndian(buffer, num);
+            output.Advance(4);
+        }
+
+        public static void WriteUInt(this IBufferWriter<byte> output, uint num)
+        {
+            var buffer = output.GetSpan(4);
+            BinaryPrimitives.WriteUInt32BigEndian(buffer, num);
+            output.Advance(4);
+        }
+
+        public static void WriteLong(this IBufferWriter<byte> output, long num)
+        {
+            var buffer = output.GetSpan(8);
+            BinaryPrimitives.WriteInt64BigEndian(buffer, num);
+            output.Advance(8);
+        }
+
+        public static void WriteULong(this IBufferWriter<byte> output, ulong num)
+        {
+            var buffer = output.GetSpan(8);
+            BinaryPrimitives.WriteUInt64BigEndian(buffer, num);
+            output.Advance(8);
+        }
+
+        public static void WriteString(this IBufferWriter<byte> output, string text)
+        {
+            var lengthBuffer = output.GetSpan(2);
+            if (string.IsNullOrEmpty(text))
+            {
+                BinaryPrimitives.WriteInt16BigEndian(lengthBuffer, 0);
+                output.Advance(2);
+                return;
+            }
+
+            var bytes = Encoding.UTF8.GetBytes(text);
```

Comment:
Here you can write straight to the span without allocating an array by doing something like

``` csharp
var textSize = Encoding.UTF8.GetByteCount(text);
output.Ensure(textSize + sizeof(short));
var span = output.Span;
BinaryPrimitives.WriteInt16BigEndian(span, (short)textSize);
span = span.Slice(sizeof(short));
Encoding.UTF8.GetBytes(text, span);
output.Advance(textSize + sizeof(short));
```

Reply:
this whole file is taken from my attempt 😄 
I've fixed mine up (I think) https://github.com/AndyPook/A6k.Kafka/blob/42e5dbf00e0e81f21cb04006c46a64f6431dd6c6/A6k.Kafka/BufferWriterExtensions.cs#L67
Feel free to check and take what you need

Reply:
Q: is it worth doing the `[MethodImpl(MethodImplOptions.AggressiveInlining)]` thing on all these extensions?

Reply:
I wouldn't to start with, maybe profile it later, if the method is small enough it will likely get inlined anyway

Reply:
what the hell (benchmarks for much later)
https://github.com/AndyPook/A6k.Kafka/blob/master/A6k.Kafka/BufferWriterExtensions.cs